### ISSUE TITLE
Focal loss calculation

### ISSUE THREAD

**Issue by speedport1502** [jcwang0602/MLLMSeg#9]
First of all, thank you for sharing your implementation.
While experimenting with the mask decoder using the QWEN-2.5VL model, I consistently observed a relatively small focal loss value:

MaskFocal 3.5320e-03 (1.0494e-02)	
MaskDICELoss 0.9461 (0.9419)

While reviewing the code, I noticed that a sigmoid is applied before returning the mask from the decoder:

score_map_seg = self.conv2(x).sigmoid()
return score_map_seg, img_vit_l_32, tgt

Later, the loss is computed using sigmoid_focal_loss, which internally calls F.binary_cross_entropy_with_logits. As far as I understand, binary_cross_entropy_with_logits expects raw logits (i.e., values before applying a sigmoid).

This made me wonder whether applying the sigmoid in the decoder is intentional in your design. Did you observe any specific benefits during training from doing this, or is there a particular reason for structuring it this way?

**Comment by jcwang0602** [comment]
Thank you for your interest in our work. 

Using the sigmoid function in the decoder is a common practice in segmentation tasks to constrain the predicted mask values between 0 and 1. Values greater than 0.5 are considered target pixels, while those below 0.5 are treated as background. 

The sigmoid_focal_loss is also configured based on previous works. 

Regarding loss calculation using raw logits, we haven't previously focused on this issue and are currently uncertain about it.


Translated with DeepL.com (free version)

**Comment by speedport1502** [comment]
Thank you for the explanation, that makes sense, and I understand the motivation for applying a sigmoid in the decoder to obtain mask probabilities in [0,1].

My question primarily concerns the focal loss implementation. In the provided focal loss implementation, the classification term is computed using F.binary_cross_entropy_with_logits from torch.nn.functional. This function is designed to operate on raw logits, as it internally combines a sigmoid activation with the binary cross-entropy computation for numerical stability (see the PyTorch documentation for binary_cross_entropy_with_logits:
https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.binary_cross_entropy_with_logits.html
 and
https://docs.pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html).

Since the decoder output also passes through a sigmoid:

score_map_seg = self.conv2(x).sigmoid()

it seems that the focal loss receives already sigmoid-normalized values rather than raw logits. From my understanding, this appears to result in the focal loss operating on sigmoid-activated outputs, with another sigmoid being applied internally by binary_cross_entropy_with_logits. I may be misunderstanding this interaction, but I was wondering whether this could be related to the relatively small focal loss values I observed.

At the same time, I noticed that the Dice loss operates directly on the provided inputs and therefore does not introduce an additional sigmoid, which makes the behavior between the two losses slightly different.

I’m unfamiliar with the concept of applying the sigmoid twice in this setting and was wondering whether this is commonly done or if it simply worked better in your experiments.

Thanks again for your clarification and for sharing the implementation.
